Readers can see editor field responses at summary page

ravimajithia opened the pull request:
- Minor text changes

Commit:
Readers can see editor field responses at summary page
- Minor text changes

otacke commented on src/scripts/summary.js:
```diff
@@ -410,17 +410,29 @@ class Summary extends H5P.EventDispatcher {
     for (const section of sections) {
       const sectionRow = document.createElement("li");
       sectionRow.classList.add('h5p-interactive-book-summary-overview-section-details');
+      const sectionRowContainer = document.createElement("div");
```
Please try not to use double quotes. I know that the ruleset of eslint here does not follow the H5P coding style guide, so your IDE will not complain - and the code styling will eventually need to get fixed anyway - but we can make it right right away :-)

ravimajithia replied:
It will be taken care of in next commit.

otacke commented on src/scripts/summary.js:
```diff
@@ -410,17 +410,29 @@ class Summary extends H5P.EventDispatcher {
+          contentChapter.classList.toggle('hidden');
+
+          // Resize the iframe
+          setTimeout(() => this.parent.trigger('resize'), 200);
```
Function bodies that are not just returning a value, even if it's one line only, should be written in braces:

```
setTimeout(() => {
  this.parent.trigger('resize');
}, 200);
```

What's the timeout needed for?

ravimajithia replied:
The reason for this is the slider that opens up on click has some transition effect on height so we need to wait when the div expanded fully and then initiate the parent resize.

otacke replied:
Then there are three better solutions at least:
a) At least add a comment what the magic number is.
b) Replace the magic number by a constant and add a comment to the constant.
c) Add a `transtitionend` listener to the `responseField` and let that one trigger the resize. Not only will you get rid of the magic number (and thus not require to change it whenever the transition time is changed) and the timeout, but also ensure that any trigger for the `responseField` to toggle its transition will trigger a resize not just a click - may never be needed here, but would be the general cover-all solution. Along with that, you may want to check the `propertyName` of the `transitionend` event though, because you used `transition all` which is a little wonky in itself to let all CSS properties transition and not just one that controls the height (there's probably some other better "accordion" solution).

ravimajithia replied:
a) I will add a comment
b) I do not think we need to define the constant for a single operation if that was repetitive I would have created an event. A comment will be sufficient enough to explain the necessity of the code.
c) I do not it is a good idea to add a transition event listener for each field we use in content type, imagine the user has 4 inputs then we are ended up adding 4 event listeners. We are just increasing application overhead by adding more event listeners. In addition, the outcome would be the same. Hence,  I would like to stick with the magic number :-)
I use transition all CSS because there is height, margin, padding and border we need to add a transition for all of them, rather than that I use transition all which seems a good option (in future if we change the design then the developer does not have to focus on the transition that's a plus)

Just a thought, if possible, please try to give all feedback at once so that we can avoid many iterations :-)
Or you can contact me directly on Slack so that we can be more effective to complete this task.

otacke replied:
Let's keep this one at least on github. It's a potential source for learning that others can use, and that's what NDLA wants. It's not common practice, but there are people who argue that not only the product should be open and transparent, but the process itself. It helps to let others understand HOW creating code works in practice instead of just seeing the result. Yes, that can mean a little more communication. @janlindso can gladly overrule me here by the way :-)

You do not need to add a `transitionend` listener/handler for every field if you're worried about performance. That's what the JavaScript gods gave us event delegation for. Events will bubble up unless propagation is stopped. You can add one single `transitionend` listener to a parent element as well, e.g. `h5p-interactive-book-summary-overview-list` and handle it there. The "outcome" will be the same, but future maintenance will be decreased. If someone wants to change the transition time later on, he/she will not have to know that this needs to be done in two places (CSS and JavaScript).

Having constants is not only useful to not repeat oneself, but to keep the code clean and to make maintenance easier. If all constants are defined in the same place, e.g. the top or bottom of the code, it's easier to find them - and you even have the benefit of naming them in order to reveal their purpose, e.g. `const responseFieldtransitionTimeMS = 200;`.

ravimajithia replied:
Thank you for explaining. No need to overrule your decision, I understood the suggested approach is more sensible for users. I have updated the code, please have a look.

otacke replied:
That should do. Good job! :+1:

If my picky mind finds something later on (the original code of Interactive Book is a little wacky anyway), I can tweak it myself ...

otacke commented on src/scripts/summary.js:
```diff
@@ -410,17 +410,29 @@ class Summary extends H5P.EventDispatcher {
       title.classList.add('h5p-interactive-book-summary-section-title');
       title.onclick = () => {
+        if (section.instance.libraryInfo.machineName === 'H5P.ActiveReaderTextInput') {
+          const contentChapter = document.querySelector(`.h5p-content-chapter-${chapterId}`);
```
Please do not use query selectors if they are not absolutely necessary. They are slow and prone to error (and the result can in fact be `null` here if you have not interacted with the CKEditor before clicking).

You have full access to the Active reader, so please retrieve the DOM from the appropriate class instance.

ravimajithia replied:
I think it is necessary, I am creating the div element on the fly for the content added in the active-reader-text-input, which does not have method to retrieve DOM. https://github.com/NDLANO/h5p-active-reader/pull/1/files#diff-44865373b09e58b12e9044bddc3a7b033ff7247ba5c3c1578d127800c87d9a3dR476

otacke replied:
I didn't say everything was there already. The `responseField` elements are created by the Active Reader (not by some other component with code out of our control) and can easily stored (e.g. in an object indexed by the `subContentId` or whatever is suitable) in order to retrieve them when needed here.

ravimajithia replied:
This could be a way of doing or achieving the same result, I prefer doing this way and it is aligned with the existing libraries' code.

otacke replied:
I'll not argue here, but why do you prefer it?

ravimajithia replied:
The reason is, even if we use the element from the active-reader-input-text library we still need to put that into DOM and in a way there will be the same outcome.

otacke replied:
I am not sure whether we're talking about the same thing here. The element that you are selecting here is not part of the `active-reader-input-text-library`. It is the `responseField` that you created here inside the summary page of the `active-reader`  (https://github.com/NDLANO/h5p-active-reader/pull/1/files#diff-44865373b09e58b12e9044bddc3a7b033ff7247ba5c3c1578d127800c87d9a3dR476) just a couple of lines below.

ravimajithia replied:
Ahh, I thought you were talking about the response method from the h5p-active-reader-text-input library. No worries we'll use the variable

ravimajithia replied:
I think this is also covered. Please have a look.

Commit:
Code improvements
- Make eslint happy
- Add height and scroll to content slider
- Some minor bug fixes

Commit:
Use variable instead queryselector to identify element

Commit:
Add resize on the transitionend events
- Add toggle indicator for users

otacke commented:
@ravimajithia Please merge yourself if you think you're done.
